### PR TITLE
Multiple activities per Agreement

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: #730
References ya-client from branch: https://github.com/golemfactory/ya-client/pull/69
Remember to change referenced commit on ya-client to commit on master branch

### CODE REVIEW

File: agent/provider/src/market/provider_market.rs [hunk 1]
```diff
@@ -385,84 +435,116 @@ async fn process_agreement(
...
+
+    loop {
+        match ctx.api.collect(&id, Some(timeout), Some(5)).await {
+            Err(error) => {
+                log::warn!("Can't query market events. Error: {}", error);
+                match error {
+                    ya_client::error::Error::HttpError { code, .. } => {
+                        if code.as_u16() == 404 {
+                            log::info!("Resubscribing subscription [{}]", id);
+                            let _ = ctx.market.send(ReSubscribe(id.clone())).await;
```
Comment (tworec) [inline agent/provider/src/market/provider_market.rs:509]:
```suggestion
                            ctx.market.do_send(ReSubscribe(id.clone()));
                            return;
```

File: agent/provider/src/market/provider_market.rs [hunk 2]
```diff
@@ -632,18 +822,46 @@ impl Handler<Unsubscribe> for ProviderMarket {
                 log::info!("Unsubscribing {} active offer(s)", subs.len());
                 subs
             }
+            OfferKind::WithIds(subs) => subs,
```
Comment (tworec) [inline agent/provider/src/market/provider_market.rs:825]: pls remove subscription from `self.subscriptions` here; same as above

File: agent/provider/src/provider_agent.rs [hunk 3]
```diff
@@ -425,6 +437,7 @@ impl Handler<CreateOffers> for ProviderAgent {
         let preset_names = match msg.0 {
             OfferKind::Any => self.presets.active(),
             OfferKind::WithPresets(names) => names,
+            OfferKind::WithIds(_) => vec![],
```
Comment (tworec) [inline agent/provider/src/provider_agent.rs:440]: should `log::warn`

File: core/market/src/negotiation/common.rs [hunk 4]
```diff
@@ -323,18 +325,26 @@ impl CommonBroker {
...
         };
+
+        terminate_reason_metric(&reason, owner_type);
         log::info!(
-            "Requestor {} terminated Agreement [{}] and sent to Provider.",
-            &id.identity,
+            "Agent {} terminated Agreement [{}]. Reason: {}",
+            &id.display(),
             &agreement.id,
+            reason.display(),
```
Comment (tworec) [inline core/market/src/negotiation/common.rs:347]: consider logging also extra field

File: core/market/src/negotiation/common.rs [hunk 5]
```diff
@@ -576,8 +599,33 @@ pub fn expect_state(
...
+}
+
+/// This function extract from Reason additional information about termination reason
+/// and increments metric counter. Note that Reason isn't required to have any fields
+/// despite 'message'.
+pub fn terminate_reason_metric(reason: &Option<Reason>, owner: OwnerType) {
+    let p_code = get_reason_code(reason, "golem.provider.code");
+    let r_code = get_reason_code(reason, "golem.requestor.code");
+
+    let reason_code = r_code.xor(p_code).unwrap_or("None".to_string());
```
Comment (tworec) [inline core/market/src/negotiation/common.rs:622]: to be consistent with ya-provider
```suggestion
    let reason_code = r_code.xor(p_code).unwrap_or("NotSpecified".to_string());
```

File: core/market/src/negotiation/provider.rs [hunk 6]
```diff
@@ -74,6 +74,9 @@ impl ProviderBroker {
         counter!("market.proposals.provider.init-negotiation", 0);
         counter!("market.agreements.provider.proposed", 0);
         counter!("market.agreements.provider.approved", 0);
+        counter!("market.agreements.provider.terminated", 0);
+        counter!("market.agreements.provider.terminated.reason", 0, "reason" => "None");
```
Comment (tworec) [inline core/market/src/negotiation/provider.rs:78]:
```suggestion
        counter!("market.agreements.provider.terminated.reason", 0, "reason" => "NotSpecified");
```

File: core/market/src/negotiation/requestor.rs [hunk 7]
```diff
@@ -84,12 +84,14 @@ impl RequestorBroker {
...
         counter!("market.agreements.requestor.rejected", 0);
         counter!("market.agreements.requestor.cancelled", 0);
-        counter!("market.agreements.requestor.terminated", 0);
         counter!("market.proposals.requestor.generated", 0);
         counter!("market.proposals.requestor.received", 0);
         counter!("market.proposals.requestor.countered", 0);
         counter!("market.events.requestor.queried", 0);
         counter!("market.agreements.events.queried", 0);
+        counter!("market.agreements.requestor.terminated", 0);
+        counter!("market.agreements.requestor.terminated.reason", 0, "reason" => "None");
```
Comment (tworec) [inline core/market/src/negotiation/requestor.rs:93]:
```suggestion
        counter!("market.agreements.requestor.terminated.reason", 0, "reason" => "NotSpecified");
```

File: core/market/src/utils/display.rs [hunk 8]
```diff
@@ -37,6 +39,12 @@ impl<'a> std::fmt::Display for DisplayEnabler<'a, Identity> {
     }
 }
 
+impl<'a> std::fmt::Display for DisplayEnabler<'a, Reason> {
```
Comment (tworec) [inline core/market/src/utils/display.rs:42]: can be removed, bc Reason has Display now

File: core/serv/src/main.rs [hunk 9]
```diff
@@ -358,6 +358,7 @@ impl ServiceCommand {
 
                     Services::rest(app, &context)
                 })
+                .keep_alive(30)
```
Comment (tworec) [inline core/serv/src/main.rs:361]:
```suggestion
                // this is maximum supported timeout for our REST API 
                .keep_alive(30)
```

File: exe-unit/Cargo.toml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ya-exe-unit"
-version = "0.1.0"
+version = "0.2.0"
```
Comment (tworec) [inline exe-unit/Cargo.toml:3]: ?
```suggestion
version = "0.1.0"
```

**Review by tworec — APPROVED** [review]

**Review by tworec — DISMISSED** [review]

File: core/market/src/utils/display.rs [hunk 11]
```diff
@@ -37,6 +39,12 @@ impl<'a> std::fmt::Display for DisplayEnabler<'a, Identity> {
...
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        write!(f, "'{}'", &self.0.message)
+    }
+}
+
```
Comment (nieznanysprawiciel) [inline core/market/src/utils/display.rs:47]:
```suggestion
```

**Review by tworec — DISMISSED** [review]

File: agent/provider/src/market/config.rs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    #[structopt(long, env, default_value = "20.0")]
+    pub agreement_events_interval: f32,
+    #[structopt(long, env, default_value = "20.0")]
+    pub negotiation_events_interval: f32,
+    #[structopt(long, env, default_value = "10.0")]
+    pub agreement_approve_timeout: f32,
+    #[structopt(long, env, default_value = "LimitAgreements")]
+    pub negotiator_type: String,
+    #[structopt(skip = "you-forgot-to-set-session-id")]
+    pub session_id: String,
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply agent/provider/src/market/config.rs:15]: As we discussed - it's not so obvious, because user could set conflicting id. To discuss later.

**Review by tworec — DISMISSED** [review]

File: agent/provider/src/market/config.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (tworec) [reply agent/provider/src/market/config.rs:15]: it should be moved upper in config as it will be used by all APIs, not only Market

**Review by tworec — APPROVED** [review]

**Review by Wiezzel — APPROVED** [review]

**Review by mfranciszkiewicz — APPROVED** [review]

**Pull request merged**